File: app/src/main/java/org/libreoffice/kit/Office.java


package org.libreoffice.kit;

import java.nio.ByteBuffer;

public class Office {
    private ByteBuffer handle;
    private MessageCallback messageCallback = null;

    public Office(ByteBuffer handle) {
        this.handle = handle;
        bindMessageCallback();
    }

    private native void bindMessageCallback();

    public native String getError();

    private native ByteBuffer documentLoadNative(String url);

    public Document documentLoad(String url) {
        ByteBuffer documentHandle = documentLoadNative(url);
        Document document = null;
        if (documentHandle != null) {
            document = new Document(documentHandle);
        }
        return document;
    }

    public native void destroy();
    public native void destroyAndExit();
    public native void setDocumentPassword(String url, String pwd);
    public native void setOptionalFeatures(long options);

    public void setMessageCallback(MessageCallback messageCallback) {
        this.messageCallback = messageCallback;
    }

    private void messageRetrievedLOKit(int signalNumber, String payload) {
        if (messageCallback != null) {
            messageCallback.messageRetrieved(signalNumber, payload);
        }

    }

    public interface MessageCallback {

        void messageRetrieved(int signalNumber, String payload);
    }
}

File: app/src/main/kotlin/org/libreoffice/kit/Office.kt
package org.libreoffice.kit

import java.nio.ByteBuffer

class Office(private val handle: ByteBuffer) {

    private var messageCallback: MessageCallback? = null

    init {
        bindMessageCallback()
    }

    private external fun bindMessageCallback()

    external fun getError(): String

    private external fun documentLoadNative(url: String): ByteBuffer?

    fun documentLoad(url: String): Document? {
        val documentHandle = documentLoadNative(url) ?: return null
        return Document(documentHandle)
    }

    external fun destroy()
    external fun destroyAndExit()
    external fun setDocumentPassword(url: String, pwd: String)
    external fun setOptionalFeatures(options: Long)

    fun setMessageCallback(messageCallback: MessageCallback?) {
        this.messageCallback = messageCallback
    }

    @Suppress("unused")
    private fun messageRetrievedLOKit(signalNumber: Int, payload: String) {
        messageCallback?.messageRetrieved(signalNumber, payload)
    }

    fun interface MessageCallback {
        fun messageRetrieved(signalNumber: Int, payload: String)
    }
}
